src/storage/src/render/multi_worker_persist_sink.rs

```diff
@@ -0,0 +1,1123 @@
+        scope,
+        collection_id,
+        operator_name.clone(),
+        &target,
+        &desired_collection.inner,
+        Arc::clone(&persist_clients),
+    );
+
+    let (written_batches, write_token) = write_batches(
+        scope,
+        collection_id.clone(),
+        operator_name.clone(),
+        &target,
+        &batch_descriptions,
+        &desired_collection.inner,
+        Arc::clone(&persist_clients),
+        storage_state,
+    );
+
+    let append_token = append_batches(
+        scope,
+        collection_id.clone(),
+        operator_name,
+        &target,
+        &batch_descriptions,
+        &written_batches,
+        persist_clients,
+        storage_state,
+        output_index,
+        metrics,
+    );
+
+    Rc::new((mint_token, write_token, append_token))
+}
+
+/// Whenever the frontier advances, this mints a new batch description (lower
+/// and upper) that writers should use for writing the next set of batches to
+/// persist.
+///
+/// Only one of the workers does this, meaning there will only be one
+/// description in the stream, even in case of multiple timely workers. Use
+/// `broadcast()` to, ahem, broadcast, the one description to all downstream
+/// write operators/workers.
+fn mint_batch_descriptions<G>(
+    scope: &mut G,
+    collection_id: GlobalId,
+    operator_name: String,
+    target: &CollectionMetadata,
+    desired_stream: &Stream<G, (Result<Row, DataflowError>, Timestamp, Diff)>,
+    persist_clients: Arc<PersistClientCache>,
+) -> (
+    Stream<G, (Antichain<Timestamp>, Antichain<Timestamp>)>,
+    Rc<dyn Any>,
+)
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let persist_location = target.persist_location.clone();
+    let shard_id = target.data_shard;
+    let target_relation_desc = target.relation_desc.clone();
+
+    // Only one worker is responsible for determining batch descriptions. All
+    // workers must write batches with the same description, to ensure that they
+    // can be combined into one batch that gets appended to Consensus state.
+    let hashed_id = collection_id.hashed();
+    let active_worker = usize::cast_from(hashed_id) % scope.peers() == scope.index();
+
+    // Only the "active" operator will mint batches. All other workers have an
+    // empty frontier. It's necessary to insert all of these into
+    // `compute_state.sink_write_frontier` below so we properly clear out
+    // default frontiers of non-active workers.
+
+    let mut mint_op = AsyncOperatorBuilder::new(
+        format!("{} mint_batch_descriptions", operator_name),
+        scope.clone(),
+    );
+
+    let (mut output, output_stream) = mint_op.new_output();
+
+    let mut desired_input = mint_op.new_input(desired_stream, Pipeline);
+
+    let shutdown_button = mint_op.build(move |mut capabilities| async move {
+        let mut cap_set = CapabilitySet::from_elem(capabilities.pop().expect("missing capability"));
```

Comment:
Can we use `into_element` from `mz_ore` or otherwise assert that there is only a single capability in there?